#! /bin/sh -e

# All lines beginning with `# DP:' are a description of the patch.
# DP: Description: Avoid infinite loops in case the mounts file is hosed.
# DP: Related bugs: 219271
# DP: Author: Ulrich Drepper
# DP: Upstream status: In CVS
# DP: Status Details: 
# DP: Date: November 5th, 2003

if [ $# -ne 2 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch) patch -d "$2" -f --no-backup-if-mismatch -p0 < $0;;
    -unpatch) patch -d "$2" -f --no-backup-if-mismatch -R -p0 < $0;;
    *)
	echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
	exit 1
esac
exit 0

# append the patch here and adjust the -p? flag in the patch calls.
--- sysdeps/unix/sysv/linux/internal_statvfs.c~	2003-11-03 18:53:20.000000000 -0500
+++ sysdeps/unix/sysv/linux/internal_statvfs.c	2003-10-28 10:24:47.000000000 -0500
@@ -17,6 +17,7 @@
    Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
    02111-1307 USA.  */
 
+#include <assert.h>
 #include <errno.h>
 #include <mntent.h>
 #include <paths.h>
@@ -167,14 +168,17 @@
 	     statvfs call got a name which was not the mount point.
 	     Check again, this time without checking for name matches
 	     first.  */
-	  if (! success)
+	  if (! success && (name != NULL || fsname != NULL))
 	    {
 	      if (name != NULL)
 		/* Try without a mount point name.  */
 		name = NULL;
-	      else if (fsname != NULL)
-		/* Try without a filesystem name.  */
-		fsname = fsname2 = NULL;
+	      else
+		{
+		  /* Try without a filesystem name.  */
+		  assert (fsname != NULL);
+		  fsname = fsname2 = NULL;
+		}
 
 	      /* It is not strictly allowed to use rewind here.  But
 		 this code is part of the implementation so it is
